server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java

```diff
@@ -0,0 +1,265 @@
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public ClearVotingConfigExclusionsResponse read(StreamInput in) throws IOException {
+                    return new ClearVotingConfigExclusionsResponse(in);
+                }
+            }
+        );
+    }
+
+    /**
+     * This method triggers batch of tasks for nodes to be decommissioned using executor {@link NodeRemovalClusterStateTaskExecutor}
+     * Once the tasks are submitted, it waits for an expected cluster state to guarantee
+     * that the expected decommissioned nodes are removed from the cluster
+     *
+     * @param nodesToBeDecommissioned set of the node to be decommissioned
+     * @param reason reason of removal
+     * @param timeout timeout for the request
+     * @param nodesRemovedListener callback for the success or failure
+     */
+    public void removeDecommissionedNodes(
+        Set<DiscoveryNode> nodesToBeDecommissioned,
+        String reason,
+        TimeValue timeout,
+        ActionListener<Void> nodesRemovedListener
+    ) {
+        final Map<NodeRemovalClusterStateTaskExecutor.Task, ClusterStateTaskListener> nodesDecommissionTasks = new LinkedHashMap<>(
+            nodesToBeDecommissioned.size()
+        );
+        nodesToBeDecommissioned.forEach(discoveryNode -> {
+            final NodeRemovalClusterStateTaskExecutor.Task task = new NodeRemovalClusterStateTaskExecutor.Task(discoveryNode, reason);
+            nodesDecommissionTasks.put(task, nodeRemovalExecutor);
+        });
+        clusterService.submitStateUpdateTasks(
+            "node-decommissioned",
+            nodesDecommissionTasks,
+            ClusterStateTaskConfig.build(Priority.URGENT),
+            nodeRemovalExecutor
+        );
+
+        Predicate<ClusterState> allDecommissionedNodesRemovedPredicate = clusterState -> {
+            Set<DiscoveryNode> intersection = Arrays.stream(clusterState.nodes().getNodes().values().toArray(DiscoveryNode.class))
+                .collect(Collectors.toSet());
+            intersection.retainAll(nodesToBeDecommissioned);
+            return intersection.size() == 0;
+        };
+
+        final ClusterStateObserver observer = new ClusterStateObserver(clusterService, timeout, logger, threadPool.getThreadContext());
+
+        observer.waitForNextChange(new ClusterStateObserver.Listener() {
+            @Override
+            public void onNewClusterState(ClusterState state) {
+                logger.info("successfully removed all decommissioned nodes [{}] from the cluster", nodesToBeDecommissioned.toString());
+                nodesRemovedListener.onResponse(null);
+            }
+
+            @Override
+            public void onClusterServiceClose() {
+                logger.warn("cluster service closed while waiting for removal of decommissioned nodes.");
```

Comment:
nit: make logging more informative here and other places

Reply:
done. Added details of nodes to be decommissioned in logs